gh-91673: Missing import in dataclasses documentation

xandrade opened the pull request:
As reported in #91673, the import is missed for the typing  ```List``` at dataclasses.rst file.

In this example:

```
  @dataclass
  class D:
      x: List = []
      def add(self, element):
          self.x += element
```
the import ```from typing import List``` is missing

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

xandrade committed:
Adding missing import in example

In this example:

```
  @dataclass
  class D:
      x: List = []
      def add(self, element):
          self.x += element
```
the import ```from typing import List``` is missing

xandrade committed:
Removed dataclasses.rst:704: trailing whitespace

JelleZijlstra commented on Doc/library/dataclasses.rst:
```diff
@@ -700,6 +700,8 @@ variable ``x``, as expected.
+  
   @dataclass
   class D:
       x: List = []
```
As @ericvsmith said on the issue, it's easier to just put `x: list = []` here.

xandrade replied:
@JelleZijlstra I think they really mean ```List```, from ```typing```. But I can be wrong. With ```x: list = []```, you get the following error:
```
ValueError: mutable default <class 'list'> for field x is not allowed: use default_factory
```

In that case, it should be (I think):

```
x: list = field(default_factory=list)
```

JelleZijlstra replied:
Using `list` is fine in recent versions; see PEP-585.

xandrade replied:
Noted, but ```x: list = []``` doesn't work either (tried in 3.7, 3.8 and 3.9). Maybe in 3.10?

```
from dataclasses import dataclass
  
@dataclass
class D:
    x: list = []
```
throws the following exception:

```---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [2], in <cell line: 4>()
      1 from dataclasses import dataclass, field
      2 from typing import List
      4 @dataclass
----> 5 class D:
      6     x: list = []

File /opt/conda/lib/python3.9/dataclasses.py:1021, in dataclass(cls, init, repr, eq, order, unsafe_hash, frozen)
   1018     return wrap
   1020 # We're called as @dataclass without parens.
-> 1021 return wrap(cls)

File /opt/conda/lib/python3.9/dataclasses.py:1013, in dataclass.<locals>.wrap(cls)
   1012 def wrap(cls):
-> 1013     return _process_class(cls, init, repr, eq, order, unsafe_hash, frozen)

File /opt/conda/lib/python3.9/dataclasses.py:863, in _process_class(cls, init, repr, eq, order, unsafe_hash, frozen)
    858 cls_annotations = cls.__dict__.get('__annotations__', {})
    860 # Now find fields in our class.  While doing so, validate some
    861 # things, and set the default values (as class attributes) where
    862 # we can.
--> 863 cls_fields = [_get_field(cls, name, type)
    864               for name, type in cls_annotations.items()]
    865 for f in cls_fields:
    866     fields[f.name] = f

File /opt/conda/lib/python3.9/dataclasses.py:863, in <listcomp>(.0)
    858 cls_annotations = cls.__dict__.get('__annotations__', {})
    860 # Now find fields in our class.  While doing so, validate some
    861 # things, and set the default values (as class attributes) where
    862 # we can.
--> 863 cls_fields = [_get_field(cls, name, type)
    864               for name, type in cls_annotations.items()]
    865 for f in cls_fields:
    866     fields[f.name] = f

File /opt/conda/lib/python3.9/dataclasses.py:747, in _get_field(cls, a_name, a_type)
    739     # Should I check for other field settings? default_factory
    740     # seems the most serious to check for.  Maybe add others.  For
    741     # example, how about init=False (or really,
   (...)
    744 
    745 # For real fields, disallow mutable defaults for known types.
    746 if f._field_type is _FIELD and isinstance(f.default, (list, dict, set)):
--> 747     raise ValueError(f'mutable default {type(f.default)} for field '
    748                      f'{f.name} is not allowed: use default_factory')
    750 return f

ValueError: mutable default <class 'list'> for field x is not allowed: use default_factory
```

JelleZijlstra replied:
The point of the example is that it doesn't work.

ericvsmith replied:
I agree with @JelleZijlstra. Since the example isn't supposed to work anyway, I think changing `List` to `list` is good enough. No other changes are needed.

xandrade replied:
Thanks guys for your time

xandrade committed:
Fixed issue on example

Updated ```x: list = []``` by ```x: list = field(default_factory=list)``` on the example to avoid error ```ValueError: mutable default <class 'list'> for field x is not allowed: use default_factory```
